### ISSUE TITLE
AI Tools are timing out

### ISSUE THREAD

**Issue by JattKisaan** [opencv/cvat#6087]
### My actions before raising this issue
- [ x] Read/searched [the docs](https://github.com/cvat-ai/cvat/tree/master#documentation)
- [ x] Searched [past issues](/issues)

I followed the basic installation steps for CVAT with serverless support, but the Segment Anything interactor buffers and then times out. The steps for installation that I followed seem identical to those that my colleague used, but he has been successful with using SAM and I have not. I tried DEXTR and a yolo detector model too, and they all time out, every time.

Then, I followed the steps that reproduce the error below exactly on an Ubuntu laptop. And it worked just fine. So I'm not sure what debugging steps to use to figure out what the difference is between my desktop and laptop. My desktop has openvpn and ssh servers installed. They aren't running in docker. I don't think this should affect anything, but I can't think of why else my computer would differ significantly from any other.

### Steps to Reproduce (for bugs)

1. Installed CVAT, using [these instructions](https://opencv.github.io/cvat/docs/administration/basics/installation/). I followed them only up to the `git clone`.
1. I downloaded `nuctl` as described [here, 2nd bullet](https://opencv.github.io/cvat/docs/administration/advanced/installation_automatic_annotation/). I created a project for `cvat` as described there, but I didn't deploy any of the models.
1. I followed [these instructions](https://www.cvat.ai/post/facebook-segment-anything-model-in-cvat) to deploy the segment anything model, i.e. this command, `cd serverless && ./deploy_cpu.sh pytorch/facebookresearch/sam/nuclio/`
1. Finally, I used the docker up command from back [here, 1st bullet](https://opencv.github.io/cvat/docs/administration/advanced/installation_automatic_annotation/) to get cvat and nuclio running.
2. I go to `localhost:8080` in Chrome, create a project and task, upload any image you'd like (I used [this](https://gardenerspath.com/wp-content/uploads/2021/05/Why-Are-My-Corn-Seedlings-Dying-Cover.jpg)).
3. Once I select Segment Anything in the interactors under AI Tools and click somewhere on the image, it buffers for 10s of seconds, and then returns with an error message about timing out. If I monitor `htop` while it is running, I see nothing happening after I click.

### Expected Behaviour

When my colleague follows similar steps, it just works;  CVAT buffers for some seconds, maybe 10 or so, and then spits out some kind of mask. No time out errors, ever. I should have a similar experience, as my hardware is much better than his.

### Current Behaviour

It simply times out,  with the error message shown in the following image
![cvat_issue](https://user-images.githubusercontent.com/101596350/235201922-954153d5-0f85-4712-8e72-54bd9ea747b1.png)

### Possible Solution

It seems to be related to the different containers not communicating with one another, judging by the logs below, though I don't know how to dig in further to find out more.

### Context

I'm trying to label 100s of images in a movie file of corn plants. SAM is invaluable for this, I really need it to work. Eventually I'd like to get it working on my 4090 GPU, but for getting things working CPU is fine for now.

### Your Environment

- Git hash commit (`git log -1`):
```
commit d026dbfc748c321b7300cf0bde3a729716c4f2bf (HEAD -> develop, origin/develop, origin/HEAD)
Author: Maria Khrustaleva <maya17grd@gmail.com>
Date:   Fri Apr 28 07:43:34 2023 +0200

    Fix issue azure.core.exceptions.ResourceExistsError: The specified blob already exists. (#6082)

```
- Docker version `docker version` (e.g. Docker 17.0.05):
23.0.5
- Are you using Docker Swarm or Kubernetes?
No
- Operating System and version (e.g. Linux, Windows, MacOS):
Ubuntu 20.04
- Code example or link to GitHub repo or gist to reproduce problem:
I hope the steps above make things clear.
- Other diagnostic information / logs:
    <details>
    <summary>Logs from `nuclio-nuclio-pth.facebookresearch.sam.vit_h` container. This one is the only one that looked significantly different between the desktop and laptop, at least to my eye. </summary>
On the desktop where things are not working, I get
```
serverless$ docker logs nuclio-nuclio-pth.facebookresearch.sam.vit_h | tail
/opt/nuclio/_nuclio_wrapper.py:395: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
/opt/nuclio/_nuclio_wrapper.py:395: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
23.04.30 15:38:52.392 sor.http.w0.python.logger (I) Init context...100% {"worker_id": "0"}
23.04.30 15:38:52.392 sor.http.w0.python.logger (D) Started
23.04.30 15:38:52.392                 processor (I) Starting event timeout watcher {"timeout": "30s"}
23.04.30 15:38:52.392 .webadmin.server.triggers (D) Registered custom route {"routeName": "triggers", "stream": false, "pattern": "/{id}/stats", "method": "GET"}
23.04.30 15:38:52.392 processor.webadmin.server (D) Registered resource {"name": "triggers"}
23.04.30 15:38:52.392                 processor (W) No metric sinks configured, metrics will not be published
23.04.30 15:38:52.392                 processor (D) Starting triggers {"triggersError": "json: unsupported value: encountered a cycle via *http.http"}
23.04.30 15:38:52.393            processor.http (I) Starting {"listenAddress": ":8080", "readBufferSize": 16384, "maxRequestBodySize": 33554432, "reduceMemoryUsage": false, "cors": null}
23.04.30 15:38:52.393 processor.webadmin.server (I) Listening {"listenAddress": ":8081"}
23.04.30 15:38:52.393                 processor (D) Processor started
```
Whereas on the laptop where things are working, I get
```
serverless$ docker logs nuclio-nuclio-pth.facebookresearch.sam.vit_h  | tail
/opt/nuclio/_nuclio_wrapper.py:395: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
/opt/nuclio/_nuclio_wrapper.py:395: DeprecationWarning: There is no current event loop
  loop = asyncio.get_event_loop()
23.04.30 15:33:17.291                 processor (I) Starting event timeout watcher {"timeout": "30s"}
23.04.30 15:33:17.293 .webadmin.server.triggers (D) Registered custom route {"routeName": "triggers", "stream": false, "pattern": "/{id}/stats", "method": "GET"}
23.04.30 15:33:17.293 processor.webadmin.server (D) Registered resource {"name": "triggers"}
23.04.30 15:33:17.294                 processor (W) No metric sinks configured, metrics will not be published
23.04.30 15:33:17.294                 processor (D) Starting triggers {"triggersError": "json: unsupported value: encountered a cycle via *http.http"}
23.04.30 15:33:17.307            processor.http (I) Starting {"listenAddress": ":8080", "readBufferSize": 16384, "maxRequestBodySize": 33554432, "reduceMemoryUsage": false, "cors": null}
23.04.30 15:33:17.310 processor.webadmin.server (I) Listening {"listenAddress": ":8081"}
23.04.30 15:33:17.310                 processor (D) Processor started
23.04.30 15:35:48.483 sor.http.w0.python.logger (I) call handler {"worker_id": "0"}
23.04.30 15:37:04.823 sor.http.w1.python.logger (I) call handler {"worker_id": "1"}
```
    </details>

**Comment by JattKisaan** [comment]
In the logs above, I noticed that on my laptop, where things are working, I can see two calls to the `call handler` which correspond to the two times I used SAM.

When I try to use SAM on my desktop, the operation times out, and there are no lines corresponding to the `call handler`. So it seems the container built from `nuclio-nuclio-pth.facebookresearch.sam.vit_h` is not communicating with the others properly? I don't know where to go from here though...

**Comment by JattKisaan** [comment]
Feel free to delete this issue. I have shifted the information here to #6041. I tried to search for related posts, but this is my first time posting a github issue, so I somehow missed that post.

**Comment by JattKisaan** [comment]
I believed that this issue was related to #6041 , however after completely removing cvat (with a `docker compose down -v` command to remove the database and all) and re-cloning the repo and starting from scratch, the problem remains. Every single AI tool times out on my desktop, but not my laptop.

I'm not sure how to go about debugging this issue, I would appreciate assistance in trying to fix this.

**Comment by bsekachev** [comment]
Hi @JattKisaan 

Try the new version of SAM from develop.


> after completely removing cvat (with a docker compose down -v command to remove the database and all

This command does not uninstall CVAT, and for sure does not remove any databases. It just stops containers. 

To upgrade to the newest version, on develop branch, you need to perform:

```docker compose -f docker-compose.yml -f docker-compose.dev.yml -f components/serverless/docker-compose.serverless.yml up -d --build```

And to re-create the serverless function with ``deploy_gpu.sh pytorch/facebookresearch/sam/nuclio/``

**Comment by adnan-earthsense** [comment]
Sorry Boris, I was unclear. I used the `docker compose down -v` just to get rid of some of the volumes that cvat creates, because those were persisting after i uninstalled and reinstalled cvat. To uninstall I deleted the cvat directory entirely, I issued `docker system purge --all`, and I used `nuctl delete functions ...` to delete any serverless functions I had registered with nuclio. Then I reinstalled everything afterwards, according to the steps outlined above.

I have pulled the latest changes and I am in the develop branch:
```
cvat$ git branch
* develop
```
I first issue 
```
docker compose -f docker-compose.yml -f docker-compose.dev.yml -f components/serverless/docker-compose.serverless.yml down
```
to bring down the containers that were launched on boot. `docker ps` returns
```
cvat$ docker ps
CONTAINER ID   IMAGE                        COMMAND                  CREATED       STATUS       PORTS     NAMES
996dac397c2c   gcr.io/iguazio/alpine:3.15   "/bin/sh -c '/bin/sl…"   4 hours ago   Up 4 hours             nuclio-local-storage-reader
```
Then I issue 
```
docker compose -f docker-compose.yml -f docker-compose.dev.yml -f components/serverless/docker-compose.serverless.yml up -d --build
```
Then I enter the `serverless` directory and issue
```
deploy_gpu.sh pytorch/facebookresearch/sam/nuclio/
```
Then I refresh the localhost:8080 and 8070 pages in google chrome. I can see the SAM AI Tool, I select it, try to use it, and again it times out in the same manner in which it has been.

**Comment by adnan-earthsense** [comment]
Also, JattKisaan is my first github account, this is my second, I'm the same person.

**Comment by JattKisaan** [comment]
Ah, figured it out. @whom-da's comment on issue #6041 referring to @yaochenglouis's comment on #2641 helped me realize the issue was possibly a firewall related one. I talked to chatgpt4 for an hour or two to learn about networking concepts and stuff, and finally was able to figure out the below solution.

Basically, because I use ssh and openvpn servers on my computer, I had to enable the firewall during that setup. `ufw` is installed but disabled on Ubuntu installations by default, apparently. Because of this the port 32772 was not open for communication, when a normal person without my servers wouldn't have to do this. So, I found out the port number with `docker ps` under the ports column for the SAM container, 32772, and I enabled it with `sudo ufw allow 32772/tcp`. Then I went to chrome, refreshed the CVAT page, and SAM worked. I assume now a similar approach will work for all models that I want to use.
